In [ ]:
from google.colab import files
files.upload()

In [2]:
!rm -r ~/.kaggle
!mkdir ~/.kaggle
!mv ./kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

rm: cannot remove '/root/.kaggle': No such file or directory


In [3]:
!kaggle competitions download -c titanic

  0% 0.00/34.1k [00:00<?, ?B/s]
100% 34.1k/34.1k [00:00<00:00, 2.34MB/s]


In [4]:
!mkdir Dataset
!cp /content/titanic.zip /content/Dataset/titanic.zip
!unzip -q /content/Dataset/titanic.zip -d /content/Dataset
!rm /content/Dataset/titanic.zip

In [5]:
!ls /content/Dataset/

gender_submission.csv  test.csv  train.csv


In [6]:
import pandas as pd
data = pd.read_csv("/content/Dataset/train.csv")
test = pd.read_csv("/content/Dataset/test.csv")
print(test.dtypes)

PassengerId      int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


In [7]:
## Feature selection:- 

## Test A:- Gender, Pcalss, Age, Parch, Fare, Embarked 
refined_data = data[["Survived", "Pclass","Sex", "Age","Fare", "Embarked"]]
refined_test = test[["Pclass", "Sex", "Age", "Fare", "Embarked"]]
print(refined_data)
##print(refined_data.dtypes)
print(refined_test.dtypes)

     Survived  Pclass     Sex   Age     Fare Embarked
0           0       3    male  22.0   7.2500        S
1           1       1  female  38.0  71.2833        C
2           1       3  female  26.0   7.9250        S
3           1       1  female  35.0  53.1000        S
4           0       3    male  35.0   8.0500        S
..        ...     ...     ...   ...      ...      ...
886         0       2    male  27.0  13.0000        S
887         1       1  female  19.0  30.0000        S
888         0       3  female   NaN  23.4500        S
889         1       1    male  26.0  30.0000        C
890         0       3    male  32.0   7.7500        Q

[891 rows x 6 columns]
Pclass        int64
Sex          object
Age         float64
Fare        float64
Embarked     object
dtype: object


In [ ]:
#m_surviours = refined_data.loc[(refined_data['Survived'] == 1) & (refined_data['Sex']=="male")]
f_survivours = refined_data.loc[(refined_data['Survived'] == 1) & (refined_data['Sex']=="female")]

print(m_surviours.shape[0])
print(f_survivours.shape[0])

109
233


In [ ]:
m_surviours.head(10)
temp1 = m_surviours.dropna(subset=['Age'])
print(temp1.head(10))
print(temp1.describe())

     Survived  Pclass   Sex    Age     Fare Embarked
21          1       2  male  34.00  13.0000        S
23          1       1  male  28.00  35.5000        S
74          1       3  male  32.00  56.4958        S
78          1       2  male   0.83  29.0000        S
81          1       3  male  29.00   9.5000        S
97          1       1  male  23.00  63.3583        C
125         1       3  male  12.00  11.2417        C
127         1       3  male  24.00   7.1417        S
146         1       3  male  27.00   7.7958        S
165         1       3  male   9.00  20.5250        S
       Survived     Pclass        Age        Fare
count      93.0  93.000000  93.000000   93.000000
mean        1.0   1.978495  27.276022   43.793865
std         0.0   0.920521  16.504803   76.649011
min         1.0   1.000000   0.420000    0.000000
25%         1.0   1.000000  18.000000    9.500000
50%         1.0   2.000000  28.000000   26.287500
75%         1.0   3.000000  36.000000   52.554200
max         1.0  

In [ ]:
temp2 = f_survivours.dropna(subset=['Age'])
print(temp2.head(10))
print(temp2.describe())

    Survived  Pclass     Sex   Age     Fare Embarked
1          1       1  female  38.0  71.2833        C
2          1       3  female  26.0   7.9250        S
3          1       1  female  35.0  53.1000        S
8          1       3  female  27.0  11.1333        S
9          1       2  female  14.0  30.0708        C
10         1       3  female   4.0  16.7000        S
11         1       1  female  58.0  26.5500        S
15         1       2  female  55.0  16.0000        S
22         1       3  female  15.0   8.0292        Q
25         1       3  female  38.0  31.3875        S
       Survived      Pclass         Age        Fare
count     197.0  197.000000  197.000000  197.000000
mean        1.0    1.822335   28.847716   55.643148
std         0.0    0.791478   14.175073   67.209168
min         1.0    1.000000    0.750000    7.225000
25%         1.0    1.000000   19.000000   13.000000
50%         1.0    2.000000   28.000000   26.250000
75%         1.0    2.000000   38.000000   77.958300
m

## Mean age among male surviours is 27.27 
## Mean age among female surviours is 26.25 


In [ ]:
m_died = refined_data.loc[(refined_data['Survived'] == 0) & (refined_data['Sex']=="male")]
f_died = refined_data.loc[(refined_data['Survived'] == 0) & (refined_data['Sex']=="female")]

print(m_died.shape[0])
print(f_died.shape[0])

468
81


In [ ]:
print(m_died.describe())
print(f_died.describe())

       Survived      Pclass         Age        Fare
count     468.0  468.000000  360.000000  468.000000
mean        0.0    2.476496   31.618056   21.960993
std         0.0    0.761411   14.056019   32.409799
min         0.0    1.000000    1.000000    0.000000
25%         0.0    2.000000   21.750000    7.795800
50%         0.0    3.000000   29.000000    9.416650
75%         0.0    3.000000   39.250000   26.000000
max         0.0    3.000000   74.000000  263.000000
       Survived     Pclass        Age        Fare
count      81.0  81.000000  64.000000   81.000000
mean        0.0   2.851852  25.046875   23.024385
std         0.0   0.450309  13.618591   24.821287
min         0.0   1.000000   2.000000    6.750000
25%         0.0   3.000000  16.750000    8.850000
50%         0.0   3.000000  24.500000   15.245800
75%         0.0   3.000000  33.250000   27.900000
max         0.0   3.000000  57.000000  151.550000


### Replacing the age with mean age and making a DNN for prediction
## Convention 
female = [0, 1]  
male = [1, 0 ] 

S, C, Q -> 1,2,3  

In [69]:
no_nadata= refined_data.dropna(subset=['Age','Fare'])
no_nadata["0"]=[0 if i=="male" else 1 for i in no_nadata["Sex"].values]
no_nadata["1"]=[1 if i=="male" else 0 for i in no_nadata["Sex"].values]
no_nadata["2"]=[1 if i=="S" else 0 for i in no_nadata["Embarked"].values]
no_nadata["3"]=[1 if i=="C" else 0 for i in no_nadata["Embarked"].values]
no_nadata["4"]=[1 if i=="Q" else 0 for i in no_nadata["Embarked"].values]

refined_test = refined_test.dropna(subset=['Age', 'Fare'])

refined_test["0"] = [0 if i=="male" else 1 for i in refined_test["Sex"].values]
refined_test["1"] = [1 if i=="male" else 0 for i in refined_test["Sex"].values]
refined_test["2"]=[1 if i=="S" else 0 for i in refined_test["Embarked"].values]
refined_test["3"]=[1 if i=="C" else 0 for i in refined_test["Embarked"].values]
refined_test["4"]=[1 if i=="Q" else 0 for i in refined_test["Embarked"].values]

## [S,C, ]
# print(no_nadata.head(5))
# print(no_nadata.dtypes)

print(refined_test.head(5))

   Pclass     Sex   Age     Fare Embarked  0  1  2  3  4
0     3.0    male  34.5   7.8292        Q  0  1  0  0  1
1     3.0  female  47.0   7.0000        S  1  0  1  0  0
2     2.0    male  62.0   9.6875        Q  0  1  0  0  1
3     3.0    male  27.0   8.6625        S  0  1  1  0  0
4     3.0  female  22.0  12.2875        S  1  0  1  0  0


<ipython-input-69-dca5d319a11d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_nadata["0"]=[0 if i=="male" else 1 for i in no_nadata["Sex"].values]
<ipython-input-69-dca5d319a11d>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_nadata["1"]=[1 if i=="male" else 0 for i in no_nadata["Sex"].values]
<ipython-input-69-dca5d319a11d>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the

In [9]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras import activations

In [75]:
model = keras.Sequential() 

## first layer
model.add(layers.Dense(128, input_shape=(8,)))
model.add(layers.BatchNormalization())
model.add(layers.Activation('relu'))
model.add(layers.Dropout(0.3))

model.add(layers.Dense(64))
model.add(layers.BatchNormalization())
model.add(layers.Activation('relu'))
model.add(layers.Dropout(0.3))

model.add(layers.Dense(32))
model.add(layers.BatchNormalization())
model.add(layers.Activation('relu'))
model.add(layers.Dropout(0.3))

model.add(layers.Dense(1)) ## 2 since either dead or alive better to give a probability
model.add(layers.Activation('sigmoid'))

In [40]:
import numpy as np
import torch

In [ ]:
print(no_nadata.dtypes)

Survived    float32
Pclass      float32
Sex          object
Age         float32
Fare        float32
Embarked     object
0           float32
1           float32
2           float32
3           float32
4           float32
dtype: object


In [70]:
no_nadata["Survived"] = np.asarray(no_nadata["Survived"]).astype(np.float32)
no_nadata["Pclass"] = np.asarray(no_nadata["Pclass"]).astype(np.float32)
no_nadata["Age"] = np.asarray(no_nadata["Age"]).astype(np.float32)
no_nadata["Fare"] = np.asarray(no_nadata["Fare"]).astype(np.float32)
no_nadata["0"] = np.asarray(no_nadata["0"]).astype(np.float32)
no_nadata["1"] = np.asarray(no_nadata["1"]).astype(np.float32)
no_nadata["2"] = np.asarray(no_nadata["2"]).astype(np.float32)
no_nadata["3"] = np.asarray(no_nadata["3"]).astype(np.float32)
no_nadata["4"] = np.asarray(no_nadata["4"]).astype(np.float32)

refined_test["Pclass"] = np.asarray(refined_test["Pclass"]).astype(np.float32)
refined_test["Age"] = np.asarray(refined_test["Age"]).astype(np.float32)
refined_test["Fare"] = np.asarray(refined_test["Fare"]).astype(np.float32)
refined_test["0"] = np.asarray(refined_test["0"]).astype(np.float32)
refined_test["1"] = np.asarray(refined_test["1"]).astype(np.float32)
refined_test["2"] = np.asarray(refined_test["2"]).astype(np.float32)
refined_test["3"] = np.asarray(refined_test["3"]).astype(np.float32)
refined_test["4"] = np.asarray(refined_test["4"]).astype(np.float32)


<ipython-input-70-113cbfdffa66>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_nadata["Survived"] = np.asarray(no_nadata["Survived"]).astype(np.float32)
<ipython-input-70-113cbfdffa66>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_nadata["Pclass"] = np.asarray(no_nadata["Pclass"]).astype(np.float32)
<ipython-input-70-113cbfdffa66>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats 

In [85]:
age_tensor = tf.convert_to_tensor(no_nadata["Age"])
survived_tensor = tf.convert_to_tensor(no_nadata["Survived"])
pclass_tensor = tf.convert_to_tensor(no_nadata["Pclass"])
fare_tensor = tf.convert_to_tensor(no_nadata["Fare"])
sex_tensor_0 = tf.convert_to_tensor(no_nadata["0"])
sex_tensor_1 = tf.convert_to_tensor(no_nadata["1"])
embark_tensor_0 = tf.convert_to_tensor(no_nadata["2"])
embark_tensor_1 = tf.convert_to_tensor(no_nadata["3"])
embark_tensor_2 = tf.convert_to_tensor(no_nadata["4"])

age_tensor_test = tf.convert_to_tensor(refined_test["Age"])
pclass_tensor_test = tf.convert_to_tensor(refined_test["Pclass"])
fare_tensor_test = tf.convert_to_tensor(refined_test["Fare"])
sex_tensor_0_test = tf.convert_to_tensor(refined_test["0"])
sex_tensor_1_test = tf.convert_to_tensor(refined_test["1"])
embark_tensor_0_test = tf.convert_to_tensor(refined_test["2"])
embark_tensor_1_test = tf.convert_to_tensor(refined_test["3"])
embark_tensor_2_test = tf.convert_to_tensor(refined_test["4"])


final_age_tensor = tf.concat([age_tensor, age_tensor_test], 0)
normalized_age_tensor = tf.linalg.normalize(final_age_tensor)[0]


final_fare_tensor = tf.concat([fare_tensor, fare_tensor_test], 0)
normalized_fare_tensor = tf.linalg.normalize(final_fare_tensor)[0]

combined_data = tf.convert_to_tensor([normalized_age_tensor[:714],pclass_tensor,normalized_fare_tensor[:714], sex_tensor_0, sex_tensor_1, embark_tensor_0, 
                                      embark_tensor_1, embark_tensor_2])
combined_outcomes = tf.convert_to_tensor(survived_tensor)


x_train = tf.transpose(combined_data[:,:600])
y_train = combined_outcomes[:600]
# y_train = tf.transpose(y_train[None, :])
y_train = tf.cast(y_train, dtype=tf.int64)
x_val = tf.transpose(combined_data[:,600:]) 
y_val = combined_outcomes[600:]
y_val = tf.cast(y_val, dtype=tf.int64)
# y_val = tf.transpose(y_val[None, :])

prediction_data = tf.convert_to_tensor([normalized_age_tensor[714:]
                                      ,pclass_tensor_test,normalized_fare_tensor[714:],sex_tensor_0_test,sex_tensor_1_test,embark_tensor_0_test,embark_tensor_1_test, 
                                      embark_tensor_2_test])
prediction_data = tf.transpose(prediction_data)

print(x_train.shape)
print(y_val[:5])

(600, 8)
tf.Tensor([0 1 1 0 0], shape=(5,), dtype=int64)


In [82]:
model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=0.002),  # Optimizer
    # Loss function to minimize
    loss=tf.keras.losses.BinaryCrossentropy(from_logits=False),
    # List of metrics to monitor
    metrics=['accuracy'],
)

history = model.fit(x_train,
    y_train,
    batch_size=15,
    epochs=20, 
    validation_data=(x_val, y_val)
    )


Epoch 1/20
40/40 [==============================] - 4s 10ms/step - loss: 0.4520 - accuracy: 0.7983 - val_loss: 0.4121 - val_accuracy: 0.7982
Epoch 2/20
40/40 [==============================] - 0s 4ms/step - loss: 0.4620 - accuracy: 0.7983 - val_loss: 0.4189 - val_accuracy: 0.7982
Epoch 3/20
40/40 [==============================] - 0s 4ms/step - loss: 0.4524 - accuracy: 0.7933 - val_loss: 0.4010 - val_accuracy: 0.8246
Epoch 4/20
40/40 [==============================] - 0s 5ms/step - loss: 0.4568 - accuracy: 0.8083 - val_loss: 0.4178 - val_accuracy: 0.8333
Epoch 5/20
40/40 [==============================] - 0s 4ms/step - loss: 0.4638 - accuracy: 0.7800 - val_loss: 0.4093 - val_accuracy: 0.8246
Epoch 6/20
40/40 [==============================] - 0s 5ms/step - loss: 0.4490 - accuracy: 0.7917 - val_loss: 0.4112 - val_accuracy: 0.8246
Epoch 7/20
40/40 [==============================] - 0s 4ms/step - loss: 0.4654 - accuracy: 0.8100 - val_loss: 0.4158 - val_accuracy: 0.8246
Epoch 8/20
40/40 [=

In [87]:
print(prediction_data.shape)
model.predict(prediction_data)

(331, 8)
11/11 [==============================] - 0s 2ms/step


array([[0.06402084],
       [0.35837883],
       [0.04892992],
       [0.13285524],
       [0.3832293 ],
       [0.15744437],
       [0.4252473 ],
       [0.16767076],
       [0.44670925],
       [0.1430059 ],
       [0.39400947],
       [0.9716254 ],
       [0.08407826],
       [0.9643603 ],
       [0.9293489 ],
       [0.06458341],
       [0.20738307],
       [0.37674442],
       [0.42911544],
       [0.37164488],
       [0.17380206],
       [0.44953793],
       [0.967923  ],
       [0.10208766],
       [0.97498935],
       [0.20488316],
       [0.40363693],
       [0.09821513],
       [0.180208  ],
       [0.37060872],
       [0.43175355],
       [0.21160072],
       [0.3860519 ],
       [0.13602474],
       [0.18294056],
       [0.11140997],
       [0.93658274],
       [0.9655971 ],
       [0.1359533 ],
       [0.40468335],
       [0.9675616 ],
       [0.36828285],
       [0.47025084],
       [0.14512268],
       [0.94294316],
       [0.9593263 ],
       [0.08655878],
       [0.120